### ISSUE TITLE
Repeated occurrence of Runge-Kutta-Nyström issues concerning their accuracy

### ISSUE THREAD

**Issue by HenryLangner** [SciML/OrdinaryDiffEq.jl#1938]
So far there where a few issues with the Runge-Kutta-Nyström solvers concerning their accuracy:

- https://github.com/SciML/DifferentialEquations.jl/issues/726
- https://github.com/SciML/DifferentialEquations.jl/issues/751
- https://github.com/SciML/OrdinaryDiffEq.jl/issues/1030
- https://github.com/SciML/OrdinaryDiffEq.jl/issues/1372

All Nyström solvers used, in these issues, are designed for 2nd order ODEs which are not dependent on the first derivative.
However most of the issues feature this type of 2nd order ODE. The lack of accuracy may just boil down to the solvers not being designed for theses types of problems. Any ideas on how to prevent these issues in the future?

**Comment by ranocha** [comment]
Thanks for listing these issues! Could you please check whether RKN methods allowing a second derivative depending on the first derivative work for the problems described there?

@chrisrackauckas Would it make sense to create a new ODE problem type for this class of problems to avoid these issues in the future? RKN methods allowing a dependency on the first derivative can continue to use the old setup but the other ones would require a new problem type without possibility to access the first derivative `du` when computing the second derivative `ddu`? Or do you have another suggestion how to handle this?

**Comment by HenryLangner** [comment]
For the testing I used the ```Nystrom4``` solver which allows a dependence on the first derivative. The code is attached.

The following table lists the difference between the numeric solution given by ```Tsit5```, ```DPRKN6``` and ```Nystrom4``` for each issue in the last time step. Since ```Nystrom4``` can only be used with fixed time steps I chose to use fixed time steps on the other solvers too, so adaptive is set false. Results for dt = 1e-04:

| Issue  | difference between ```Tsit5``` - ```Nystrom4```  | difference between ```Tsit5``` - ```DPRKN6``` |
| ------------- | ------------- |------------- |
| 726 | dx = 3.552713678800501e-15 | dx = 0.00014262397589970277 |
| 726 | dy = -1.4210854715202004e-14  |dy = 4.2214684228270016e-5|
| 726 | x = 3.501554601825774e-11  | x =  0.035219797821355314| 
| 726 | y =  -8.003553375601768e-11| y = 0.021683669002868555 |
||||
| 751 | u = 2.673914423212409e-10  | u = 5.44898921289132 | 
| 751 | du = 1.2369127944111824e-10 | du = 2.4798490935190785 | 
||||
| 1030 | dx = 7.086646831623969e-5 | dx = -0.49883487250072506|
| 1030 | dy = 1.388565170623579e-7  | dy =  -0.0018123043314393072|
| 1030 | x = 2.1609006668477093e-5 | x = -1.5611792619641611| 
| 1030 | y = 4.349838630445598e-7| y = -0.07033738943846785 |
||||
| 1372 | dx = 3.552713678800501e-15 | dx = -1.3117096002694987e-5|
| 1372| dy = -1.4210854715202004e-14  | dy =  -7.0716174730312e-5|
| 1372 | x = 3.501554601825774e-11  | x = 0.00011958319965943431| 
| 1372| y =  -8.003553375601768e-11| y = 0.009709477552178214 |

As seen above the solutions from ```Tsit5``` and ```Nystrom4``` are really close together, so ```Nystrom4``` works fine. 

In https://github.com/SciML/OrdinaryDiffEq.jl/issues/1372 and in https://github.com/SciML/DifferentialEquations.jl/issues/751 ```DPRKN6``` with adaptive step size gives even worse solutions. From which we can conclude that fixed time steps (```dt = 1e-04```) favor accuracy in our case (at least for tolerances under 1e-12) That being said it is odd that ¸```DPRKN6```gives different results than ```Tsit5``` and ```Nystrom4``` do. 

Digging deeper into it I find that plenty of the implemented Runge-Kutta-Nyström solvers, which are constructed for 2nd order ODEs without dependence on the first derivative, are inaccurate and show a similar error. Taking https://github.com/SciML/DifferentialEquations.jl/issues/751 for example. The analytical solution in the second argument yields 134.50945587649028 at t = 3.0. Solving it numerically  (again with fixed timesteps ```dt=1e-04```) with different RKN solvers we get

| Solver  | numerical solution at ```t=3.0```  | error at ```t=3.0``` |
| ------------- | ------------- |------------- |
| ```Nystrom4``` | 134.50945587597616 | 5.1412e-10  |
| | | |
| ```ERKN4``` | 134.4840611591785 | 0.025394717311769455  |
| ```ERKN5``` | 134.48406115917854 | 0.025394717311741033 |
| ```ERKN7``` | 134.48406115917854 | 0.025394717311741033 |
| | | |
| ```Nystrom4Velocity Independent``` | 134.4840611591781 | 0.02539471731216736  |
| ```Nystrom5Velocity Independent``` | 134.48406115917854 |0.025394717311741033  |
| | | |
| ```DPRKN4``` | 134.4840611591785  | 0.025394717311769455  |
| ```DPRKN5``` | 134.48406162812904 |0.025394248361237715  |
| ```DPRKN6``` | 134.4819611976302 | 0.027494678860080057  |
| ```DPRKN6FM``` | 134.48406115917854 | 0.025394717311741033 |
| ```DPRKN8``` |  134.4840611591785 |0.025394717311769455 |
| ```DPRKN12``` | 134.48406115917854  | 0.025394717311741033  |

The error is the difference between the numerical and analytical solution at ```t=3.0```
Note that ```Nystrom4``` is the only solver which allows the 2nd order ODE to depend on the first derivative. All the other solvers listed are designed for first derivative independent 2nd order ODEs.  

In comparison to ```Nystrom4``` all other solvers are similarly inaccurate which indicates that it is not a ```DPRKN6``` (or as seen in https://github.com/SciML/OrdinaryDiffEq.jl/issues/1030 a ```DPRKN12```) issue rather that these solvers in general have trouble solving 2nd order ODEs which depend on the first derivative.  

The code used, it is just slightly changed from the Issues listed:
[test_726.txt](https://github.com/SciML/OrdinaryDiffEq.jl/files/11352189/test_726.txt)
[test_751.txt](https://github.com/SciML/OrdinaryDiffEq.jl/files/11352190/test_751.txt)
[test_1030.txt](https://github.com/SciML/OrdinaryDiffEq.jl/files/11352191/test_1030.txt)
[test_1372.txt](https://github.com/SciML/OrdinaryDiffEq.jl/files/11352192/test_1372.txt)

**Comment by ranocha** [comment]
Thanks a lot  for these tests! This really tells us that it may be too easy for users to accidentally use the wrong method. Let's see what @ChrisRackauckas (and other main developers) think about this.

@HenryLangner Could you please comment briefly in the related issues, explaining the problem briefly and pointing to this issues? This should make it easier for new users running into the same problems to identify the cause.

**Comment by ChrisRackauckas** [comment]
Yeah this problem is tricky. The reason why we've kept these a bit open before is because they are still convergent, just at a lower order. But that's true of all sorts of methods. With highly stiff equations for example, you have the phenomena of stage order, where for example most ESDIRK methods converge at order 2 in such cases even though they are nominally orders 2-5. Similarly in the SDE case, where if you use a method for additive noise in a non-additive noise case you get order 0.5 convergence instead of the order 1.5.

We've tried to make sure that these are documented as much as possible with the solvers (https://docs.sciml.ai/DiffEqDocs/stable/solvers/dynamical_solve/#Runge-Kutta-Nystr%C3%B6m-Integrators), though I can see how "Allows acceleration to depend on velocity." slipped through telling people that others don't allow it for higher order convergence. The methods still achieve first order convergence of course, and that means they are bad.

So what to do? 

> @ChrisRackauckas Would it make sense to create a new ODE problem type for this class of problems to avoid these issues in the future? RKN methods allowing a dependency on the first derivative can continue to use the old setup but the other ones would require a new problem type without possibility to access the first derivative du when computing the second derivative ddu? Or do you have another suggestion how to handle this?

We've tried to avoid that because it can be very difficult to support so many problem types hanging around. One easy thing that we could do is do a single call to the acceleration function with velocity set as NaN values, and then error with these methods if the derivative is NaN. That of course can have some weird caveats though. I was hoping to someday use compiler plugin tools and do these kinds of proofs via abstract interpretation, but you can then only do that to Julia codes so it's limiting. I have though about doing this kind of stuff at problem construction time and tagging traits for things like this, and is_autonomous (which would allow Rosenbrock methods to simplify), etc.

But maybe doing the new problem type is easiest. What would it be named? `VelocityFreeSecondOrderODEProblem`?

**Comment by ranocha** [comment]
> We've tried to make sure that these are documented as much as possible with the solvers (https://docs.sciml.ai/DiffEqDocs/stable/solvers/dynamical_solve/#Runge-Kutta-Nystr%C3%B6m-Integrators), though I can see how "Allows acceleration to depend on velocity." slipped through telling people that others don't allow it for higher order convergence. The methods still achieve first order convergence of course, and that means they are bad.

I guess a first step could be to add explicit warnings that the second derivative must not depend on the first derivative to the docs, following the docstring updates @HenryLangner did

**Comment by ranocha** [comment]
Adding a type parameter to `DynamicalODEProblem` and `SecondOrderODEProblem` may be okay - either assuming that the second derivative depends on the first derivative by default or trying to check it as you described (for floating point types). This parameter could also be used to improve the efficiency of some methods since less computations need to be performed.

**Comment by julianmuehlenhoff** [comment]
Hi there,

I did a comparison of lots of different solvers for a second order problem and just wanted to contribute to you with another view on these solvers. For my research work, I have to find numerical solutions of very smooth second order ODE's, where also the acceleration does *not* depend on velocity. The vector-valued ODEs have the general shape of $\ddot{x}_i = \sum_j \sin (x_i - x_j)$ with some additional constants I omitted here. Since I have to calculate a massive amount of different parameters, I did research on the best fitting solver for my problem w.r.t. calculation time. For the given julia and package versions below, I did not found problems as you described above, maybe some of them were already fixed? Maybe also my problem is too smooth and easy, as it requires only a low amount of total steps with good initial guesses for them. Since this issue was never closed, I was sceptical about these DPRKN solver and just wanted to feed my own experience back to the public.

Below you can see graphs of the number of functions evaluations versus the reached global error (not a relative or absolute local error) for `adaptive = false` and different constant time steps `dt` with `Float64` numbers (hence the cutoff at around `1.0e-12`). As you can see, the DPRKN methods in general are fine for my problem with the only exception of `DPRKN5`. Maybe there is something still wrong with `DPRKN5`?

[solvers_order1_nfev.pdf](https://github.com/user-attachments/files/20865732/solvers_order1_nfev.pdf)
[solvers_order2_nfev.pdf](https://github.com/user-attachments/files/20865731/solvers_order2_nfev.pdf)

* `versioninfo()`:
```
Julia Version 1.11.5
Commit 760b2e5b73 (2025-04-14 06:53 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 12 × AMD Ryzen 5 3600XT 6-Core Processor
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, znver2)
Threads: 1 default, 0 interactive, 1 GC (on 12 virtual cores)
```

* `Pkg.status()`:
```
  [6e4b80f9] BenchmarkTools v1.6.0
  [0c46a032] DifferentialEquations v7.16.1
  [1ed8b502] SciMLSensitivity v7.85.0
  [05bca326] SimpleDiffEq v1.11.1
  [90137ffa] StaticArrays v1.9.13
```

**Comment by ChrisRackauckas** [comment]
Can you run the convergence test on that problem?

**Comment by julianmuehlenhoff** [comment]
> Can you run the convergence test on that problem?

Sorry, which convergence test on which problem?

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/DiffEqDevTools.jl/blob/master/src/convergence.jl#L109C10-L109C39 analyticless_test_convergence with the ODEProblem you see is an issue, and see if it's not converging at 5th order.

**Comment by julianmuehlenhoff** [comment]
I now did run the `DiffEqTools.analyticless_test_convergence()` on my problem. Since I am not an experienced programmer nor very good with numerics, you have to be a bit careful with my results.

I have used the following parameters/setup for `DiffEqTools.analyticless_test_convergence()`:
```
dts = 1 ./(2 .^ (12:-1:1)) # Array of time steps to test
test_setup = Dict( # The setup for the "close-to-exact" numerical solution
  :alg    => DPRKN12(), # Most accurate second order solver for my problem
  :dt     => 1.0e-4,    # Time step which is by far sufficient for my problem
  :adaptive => false,   # Adaptivity not needed for my problem
)
test = DiffEqDevTools.analyticless_test_convergence(dts, prob, solver_to_be_tested, test_setup)
plot(test)
```
There was no documentation on `DiffEqDevTools.analyticless_test_convergence`; so if you wanted me to test something differently, let me know!

I tested the following 3 solvers:
* `solver_to_be_tested = DPRKN4()`
![Image](https://github.com/user-attachments/assets/1895e945-827c-44c1-b74c-5bd06e860d27)

* `solver_to_be_tested = DPRKN5()` (the possibly problematic one)
![Image](https://github.com/user-attachments/assets/cb310012-f0f6-487d-8b9b-3c069d57af0c)
(For better understanding: L∞ and l∞ are the same)

* `solver_to_be_tested = DPRKN6()`
![Image](https://github.com/user-attachments/assets/f5ce869b-3790-4c73-a347-59af38eba3d7)

I don't know if this will help you. Alternatively, I can offer you the julia source code of my SecondOrderODEProblem including ODE functions, parameters and so on, but it will not be as tiny as a typically MWE.

**Comment by ChrisRackauckas** [comment]
Yes that helps a lot. It's saying DPRKN5 is only converging at 1st order on that problem. Is it autonomous?

**Comment by julianmuehlenhoff** [comment]
The ODE is autonomous (i.e. it does not depend on time). The general shape of the ODE – excluding some constants – is:

$\ddot{x}_i = \sum_j \sin \left( x_i - x_j \right); \ddot{x}_i \neq f \left( \dot{x}_j \right) \forall j$

**Comment by ChrisRackauckas-Claude** [comment]
I've identified and fixed the DPRKN5 convergence bug. The root cause was a copy-paste error in the CompiledFloats version of the coefficient tableau.

**The Bug:**

In `lib/OrdinaryDiffEqRKN/src/rkn_tableaus.jl`, the CompiledFloats version of `DPRKN5ConstantCache` had the wrong coefficients for stage 5:

```julia
# Wrong (was using a61 values):
a51 = convert(T, 7 // 90)   # 0.0777...
a53 = convert(T, 4 // 15)   # 0.2666...
a54 = convert(T, 1 // 15)   # 0.0666...

# Correct (from rational version):
a51 = convert(T, 0.0703125) # 9/128
a53 = convert(T, 0.140625)  # 9/64
a54 = convert(T, 0.0703125) # 9/128
```

**Verification:**

For RKN methods, each row sum must satisfy: `sum(a_ij) = c_i²/2`

For stage 5 with c4 = 3/4:
- Expected: (3/4)²/2 = 9/32 = 0.28125
- Wrong coefficients: 7/90 + 4/15 + 1/15 = 37/90 ≈ 0.411 ❌
- Correct coefficients: 9/128 + 9/64 + 9/128 = 9/32 = 0.28125 ✓

**Impact:**

This explains why DPRKN5 showed only ~1-2nd order convergence with Float64 in @julianmuehlenhoff's tests, while DPRKN4 and DPRKN6 (and DPRKN5 with BigFloat using the correct rational coefficients) achieved their expected convergence orders.

Fix: https://github.com/SciML/OrdinaryDiffEq.jl/pull/2948
